### ISSUE TITLE
Why Optuna getting stuck after certain number of trials

### ISSUE THREAD

**Issue by mostafiz67** [optuna/optuna#2820]
I am trying to do hypertuning using `Optuna`. The dataset is the `MovieLense` (1M). In one script I have `Lasso`, `Ridge` and `Knn`. Optuna is working fine for the Lasso and Ridge but getting stuck for the Knn. 

You can see the trials for the Ridge model tuning was done at `2021-07-22 18:33:53`. Later a new study was created for the Knn at `2021-07-22 18:33:53`. Now (at the time of posting) it is `2021-07-23 11:07:48` but there was no trial for the `Knn`.   

```
^[[32m[I 2021-07-22 18:33:53,959]^[[0m Trial 199 finished with value: -1.1917496039282074 and parameters: {'alpha': 3.553292157377711e-07, 'solver': 'sag', 'normalize': False}. Best is trial 71 with value: -1.1917485424789929.^[[0m
^[[32m[I 2021-07-22 18:33:53,961]^[[0m A new study created in memory with name: no-name-208652b3-68ec-4464-a2ae-5afefa9bf133^[[0m
```

The same thing is happening with the `SVR` model (you can see optuna stuck after 84 number trial at `2021-07-23 05:13:40`)
```
^[[32m[I 2021-07-23 05:13:37,907]^[[0m Trial 83 finished with value: -1.593471166487258 and parameters: {'C': 834.9834466420455, 'epsilon': 99.19181748590665, 'kernel': 'linear', 'norm': 'minmax'}. Best is trial 61 with value: -1.553044709891868.^[[0m
^[[32m[I 2021-07-23 05:13:40,261]^[[0m Trial 84 finished with value: -1.593471166487258 and parameters: {'C': 431.4022584640214, 'epsilon': 2.581688694428477, 'kernel': 'linear', 'norm': 'minmax'}. Best is trial 61 with value: -1.553044709891868.^[[0m
```
Could you tell me why Optuna is getting stuck and how can I solve the issues?

### Environment

- Optuna version: 2.8.0
- Python version: 3.8
- OS: Linux CentOS 7
- (Optional) Other libraries and their versions: Scikit Learn, Pandas, and (most common libraries)

### Reproducible examples

The code I am using for hypertuning

```
def tune(objective):
    study = optuna.create_study(direction="maximize")
    study.optimize(objective, n_trials=200, n_jobs=40)
    params = study.best_params
    return params

def knn_objective(X_train: DataFrame, y_train: DataFrame, cv_method: kfolds) -> Callable[[Trial], float]:
    def objective(trial: Trial) -> float:
        args: Dict = dict(
            n_neighbors=trial.suggest_int("n_neighbors", 2, 40, 1),
            weights=trial.suggest_categorical("weights", ["uniform", "distance"]),
            metric=trial.suggest_categorical("metric", ["euclidean", "manhattan", "mahalanobis"]),
        )
        estimator = KNeighborsRegressor(**args)
        scores = cross_validate(
            estimator, X=X_train, y=y_train, scoring="neg_mean_squared_error", cv=cv_method, n_jobs=-1
        )
        return float(np.mean(scores["test_score"]))

    return objective
```

**Comment by nzw0301** [comment]
Could you share the reproducible code?

**Comment by nzw0301** [comment]
Unfortunately, we could not get further information for investigation for a while. Let me close this issue, but please feel free to reopen if you give the reproducible code.

**Comment by Onypony** [comment]
I also have experienced this several times, the trial seems to be stuck at an iteration but no error log appears.

**Comment by nzw0301** [comment]
Could you share the reproducible code? Otherwise, we cannot investigate it. One possibility is that we might draw a very bad hyper-parameter combination that takes a too long time to evaluate objective value depending on the hyper-parameter space.

**Comment by Onypony** [comment]
I cannot even reproduce the error myself, because it does not happen always. Sometimes the training continues for 5 trials before it encounters this issue. Other times it happens in the second or the third trial. I don't know what values it selects that causes this.

But I guess what you said is possible, I thought of that too. I suppose the problem occurs when it tries to select a learning rate between 1e-3 and 1e-2, the weight decay between 1e-4 and 1e-3. It selects values with 12 non-decimal digits! Not only it slows down the computation but it also increases the number of possibilities exponentially. Do you have any way to limit the number of floating point digits? I don't think that even using step size can be helpful in this case.

**Comment by nzw0301** [comment]
During such first several trials, performance degradation might not happen in practice due to Optuna in practice. (If # trials is huge, it might happen due to TPE or sampling algorithm's time complexity.) So the issue comes from the definition of objective function I suppose. Again, we find really difficulty the cause of the issue without code. If your code is not reproducible, it would be fine to give such a code.

**Comment by sakimarquis** [comment]
It seems the problem with n_jobs != 1. I encountered the same problem with a relatively low probability, and never experience it again after not specifying n_jobs. And if someone can still reproduce this problem, I suggest you could profile your codes (with line profiler maybe).

**Comment by LegendaryKim** [comment]
Even though I set n_jobs =1, I found same issue. GPU/CPU is using but, it doese not return any finished result.
<img width="699" alt="Screen Shot 2022-01-19 at 4 11 52 PM" src="https://user-images.githubusercontent.com/13066536/150214417-29d5200e-8f1b-4b88-88d1-c4dfd7a9b1c7.png">

**Comment by tzvictor** [comment]
I had the same problem when trying to tune the SVR using Optuna. I think it may caused by the hyperparameter choosing process, when I just use the rbf kernel, it works just fine. But the study process got stuck whenever i use two kernels or more.

**Comment by nzw0301** [comment]
@tzvictor Could you share the reproducible code? Otherwise, we cannot investigate it. We might draw a very bad hyper-parameter combination that takes too long to evaluate objective value depending on the hyper-parameter space.

**Comment by tzvictor** [comment]
> @tzvictor Could you share the minimal reproducible code? Otherwise, we cannot investigate it. We might draw a very bad hyper-parameter combination that takes too long to evaluate objective value depending on the hyper-parameter space.

Sure, the code is below.  Just add one more kernel in the objective function could cause the stuck, like 'linear'.

import os

def objective(trial):
    """Define the objective function"""

    params = {
        'C': trial.suggest_loguniform('C', 0.001, 100),
        'gamma': trial.suggest_loguniform('gamma', 0.0001, 100),
        'epsilon': trial.suggest_uniform('epsilon', 0.0001, 100),
        'kernel':trial.suggest_categorical('kernel',['rbf'])
    }

    # Fit the model
    optuna_model = SVR(**params)
    optuna_model.fit(X_train, y_train)

    # Make predictions
    y_pred = optuna_model.predict(X_test)

    # Evaluate predictions
    rmse = MSE(y_test, y_pred)**0.5
    return rmse

**Comment by nzw0301** [comment]
@tzvictor Reproducible means that we get the same result if we run the provided code. So I'm afraid the code is not reproducible because the code doesn't include dataset, optuna parameters etc. Thus we cannot investigate the issue

**Comment by Moloch0** [comment]
I had the same problem with ddp using lightning as you. Only studies created for the first time can continue to be optimized, then cuda initialization fails after a few trials, and the study becomes unusable. Usage of that crushed study causes cuda initialization fails. I can't repeat it very well, because it never knows when it will fail...
